Unreadable panic message for excessively large varint

I was adjusting some error codes in my application, and I suddenly received this panic:

```
panic: (struct { message string; num uint64 }) 0x140002a4150
```

It was intermittent (due to the nature of my particular application error), and I was confused about the actual message for a bit before I finally investigated the stack trace to see that it was this panic in the qvarint package: https://github.com/quic-go/quic-go/blob/756bdc0104a1f8375eae5a1ac5a88cd7b318ba55/quicvarint/varint.go#L167-L172

I addressed the issue by ensuring my application or stream error codes fit into 62 bits. (I was setting my error constants to randomly generated 64-bit values, so it was likely that at least one of bits 63 or 64 would be set.)

Here is a minimal reproducer of the error:

```go
package main

import (
	"context"
	"fmt"
	"time"

	"github.com/madflojo/testcerts"
	"github.com/quic-go/quic-go"
)

func main() {
	ca := testcerts.NewCA()
	hostCerts, err := ca.NewKeyPair("localhost")
	if err != nil {
		panic(err)
	}

	hostTLSConfig, err := hostCerts.ConfigureTLSConfig(ca.GenerateTLSConfig())
	if err != nil {
		panic(err)
	}

	ln, err := quic.ListenAddr("localhost:8888", hostTLSConfig, nil)
	if err != nil {
		panic(err)
	}
	defer ln.Close()

	clientCerts, err := ca.NewKeyPair("localhost")
	if err != nil {
		panic(err)
	}
	clientTLSConfig, err := clientCerts.ConfigureTLSConfig(ca.GenerateTLSConfig())
	if err != nil {
		panic(err)
	}

	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	outboundConn, err := quic.DialAddr(ctx, "localhost:8888", clientTLSConfig, nil)
	if err != nil {
		panic(err)
	}

	// Accept the connection on the host.
	if _, err := ln.Accept(ctx); err != nil {
		panic(err)
	}

	s, err := outboundConn.OpenStream()
	if err != nil {
		panic(err)
	}

	if _, err := s.Write([]byte("test")); err != nil {
		panic(err)
	}

	fmt.Println("About to cancel write stream")
	// This value exceeds the 1<<62 limit on var ints,
	// and the resulting panic is unreadable:
	/*
		panic: (struct { message string; num uint64 }) 0x14000223050
	*/
	s.CancelWrite(1 << 63)

	// Ensure the process stays alive long enough for the cancel to be processed.
	time.Sleep(time.Second)
}

```

The resulting output begins like:
```console
$ go run main.go
About to cancel write stream
panic: (struct { message string; num uint64 }) 0x140002a4150

goroutine 46 [running]:
github.com/quic-go/quic-go/quicvarint.Len(...)
```

It is clear that the intent of the panic was to display the message `"value doesn't fit into 62 bits"`. If it did display that, it would have saved me a measurable amount of debugging time. It appears that if the type passed to panic has a `String() string` method, that string value will be rendered instead: https://go.dev/play/p/8KbwnmCmVdO

--- a/quicvarint/varint.go
+++ b/quicvarint/varint.go
@@ -20,6 +20,14 @@ const (
 	maxVarInt8 = 4611686018427387903
 )
 
+type varintLengthError struct {
+	Num uint64
+}
+
+func (e *varintLengthError) Error() string {
+	return fmt.Sprintf("value doesn't fit into 62 bits: %d", e.Num)
+}
+
 // Read reads a number in the QUIC varint format from r.
 func Read(r io.ByteReader) (uint64, error) {
 	firstByte, err := r.ReadByte()
@@ -118,7 +126,7 @@ func Append(b []byte, i uint64) []byte {
 			uint8(i >> 24), uint8(i >> 16), uint8(i >> 8), uint8(i),
 		}...)
 	}
-	panic(fmt.Sprintf("%#x doesn't fit into 62 bits", i))
+	panic(&varintLengthError{Num: i})
 }
 
 // AppendWithLen append i in the QUIC varint format with the desired length.
@@ -168,8 +176,5 @@ func Len(i uint64) int {
 	}
 	// Don't use a fmt.Sprintf here to format the error message.
 	// The function would then exceed the inlining budget.
-	panic(struct {
-		message string
-		num     uint64
-	}{"value doesn't fit into 62 bits: ", i})
+	panic(&varintLengthError{Num: i})
 }

--- a/quicvarint/varint_test.go
+++ b/quicvarint/varint_test.go
@@ -2,6 +2,7 @@ package quicvarint
 
 import (
 	"bytes"
+	"fmt"
 	"io"
 	"math/rand/v2"
 	"testing"
@@ -125,7 +126,10 @@ func TestVarintEncoding(t *testing.T) {
 	}
 
 	t.Run("panics when given a too large number (> 62 bit)", func(t *testing.T) {
-		require.Panics(t, func() { Append(nil, maxVarInt8+1) })
+		require.PanicsWithError(t,
+			fmt.Sprintf("value doesn't fit into 62 bits: %d", maxVarInt8+1),
+			func() { Append(nil, maxVarInt8+1) },
+		)
 	})
 }
 
@@ -203,7 +207,10 @@ func TestLen(t *testing.T) {
 	}
 
 	t.Run("panics on too large number", func(t *testing.T) {
-		require.Panics(t, func() { Len(maxVarInt8 + 1) })
+		require.PanicsWithError(t,
+			fmt.Sprintf("value doesn't fit into 62 bits: %d", maxVarInt8+1),
+			func() { Len(maxVarInt8 + 1) },
+		)
 	})
 }
 

